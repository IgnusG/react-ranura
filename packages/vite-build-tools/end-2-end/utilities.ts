import fs from "node:fs";
import path from "node:path";
import url from "node:url";

export const __dirname = url.fileURLToPath(new URL(".", import.meta.url));

export function readFile(relativePath: string) {
	return fs.readFileSync(path.join(__dirname, relativePath)).toString();
}

export function cleanup() {
	fs.writeFileSync(
		path.join(__dirname, "fixtures/package.json"),
		JSON.stringify({ name: "fixtures" }, undefined, 2),
	);

	fs.rmSync(path.join(__dirname, "fixtures/dist"), { force: true, recursive: true });
}
