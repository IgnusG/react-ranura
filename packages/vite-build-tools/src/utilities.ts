import path from "node:path";
import { cwd } from "node:process";

import { ResolvedConfig, UserConfig } from "vite";

import { Entry, EntryOptions, PluginOptions } from "./types.js";

export function getPackageJSONPath({ config }: { config: ResolvedConfig | UserConfig }) {
	return path.join(config.root ?? cwd(), "package.json");
}

export function addRelativeDot(path: string) {
	if (path.startsWith("./")) return path;

	return `./${path}`;
}

export function removeRelativeDot(path: string) {
	if (!path.startsWith("./")) return path;

	return path.replace(/^\.\//, "");
}

/** Normalize input entry options and their paths into a standardized format
 *
 * sourcePath: absolute path to the entry's source file
 *
 * - outputPath:
 * 	   - if `entryOptions.outputPath` not provided an extensionless path based on the sourcePath relative to the root directory
 *     - if provided must be a relative extensionless path
 *
 * - exportPath:
 * 	   - if `entryOptions.exports.isMain` is true, "."
 *     - if `entryOptions.exports.exportPath` not provided the `./${outputPath}`
 *     - if provided must be a relative extensionless path
 */
export function createReduceEntryOptionsToEntries({
	config,
	options,
}: {
	config: UserConfig;
	options: PluginOptions;
}) {
	return function reduceEntryOptionsToEntries(
		result: Record<string, Entry>,
		entryOptions: EntryOptions,
		index: number,
	): Record<string, Entry> {
		const sourcePath = path.isAbsolute(entryOptions.sourcePath)
			? entryOptions.sourcePath
			: path.join(config.root ?? cwd(), entryOptions.sourcePath);

		const sourcePathChunks = path.parse(sourcePath);

		const outputPath =
			entryOptions.outputPath ??
			path.join(
				path.relative(
					path.join(config.root ?? cwd(), options.sourceRoot ?? ""),
					sourcePathChunks.dir,
				),
				sourcePathChunks.name,
			);

		const exportPath = entryOptions.exports?.exportPath ?? outputPath;

		const entry: Entry = {
			sourcePath,
			outputPath: removeRelativeDot(outputPath),
			exportPath: entryOptions.exports?.isMain ? "." : addRelativeDot(exportPath),
		};

		return {
			...result,
			[index.toString()]: entry,
		};
	};
}
