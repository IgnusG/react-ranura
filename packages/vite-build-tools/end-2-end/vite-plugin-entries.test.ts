import fs from "node:fs";
import path from "node:path";

import { build, InlineConfig } from "vite";
import { afterAll, beforeEach, describe, expect, test } from "vitest";

import entriesPlugin, { PluginOptions } from "../src/vite-plugin-entries.js";
import { __dirname, cleanup, readFile } from "./utilities.js";

const defaultBuidConfig: InlineConfig = {
	root: path.join(__dirname, "fixtures"),
	build: {
		target: ["node18"],
		minify: false,
	},
};

describe("vite-plugin-entries", () => {
	beforeEach(() => {
		cleanup();
	});

	afterAll(() => {
		cleanup();
	});

	test("correctly creates the output build files for a simple es configuration", async () => {
		const buildConfig: PluginOptions = {
			formats: ["es"],
			entries: [{ sourcePath: "index.ts" }],
		};

		await build({
			...structuredClone(defaultBuidConfig),
			plugins: [entriesPlugin(buildConfig)],
		});

		expect(readFile("fixtures/dist/index-06b5571c.es.js")).toMatchSnapshot();

		expect(JSON.parse(readFile("fixtures/package.json"))).toEqual(
			expect.objectContaining({
				exports: {
					"./index": {
						import: "./dist/index-06b5571c.es.js",
						default: "./dist/index-06b5571c.es.js",
					},
				},
			}),
		);
	});

	test("correctly creates the output build files for a es/cjs mixed configuration", async () => {
		const buildConfig: PluginOptions = {
			formats: ["es", "cjs"],
			entries: [{ sourcePath: path.join(__dirname, "fixtures/index.ts") }],
		};

		await build({
			...structuredClone(defaultBuidConfig),
			plugins: [entriesPlugin(buildConfig)],
		});

		expect(readFile("fixtures/dist/index-06b5571c.es.js")).toMatchSnapshot();
		expect(readFile("fixtures/dist/cjs/index-6fade66e.cjs.js")).toMatchSnapshot();

		expect(JSON.parse(readFile("fixtures/dist/cjs/package.json"))).toEqual(
			expect.objectContaining({
				type: "commonjs",
			}),
		);

		expect(JSON.parse(readFile("fixtures/package.json"))).toEqual(
			expect.objectContaining({
				exports: {
					"./index": {
						import: "./dist/index-06b5571c.es.js",
						require: "./dist/cjs/index-6fade66e.cjs.js",
						default: "./dist/index-06b5571c.es.js",
					},
				},
			}),
		);
	});

	test("correctly creates the output build files for a es/cjs mixed configuration for multiple entries", async () => {
		const buildConfig: PluginOptions = {
			formats: ["es", "cjs"],
			entries: [
				{ sourcePath: path.join(__dirname, "fixtures/index.ts") },
				{ sourcePath: path.join(__dirname, "fixtures/other.ts") },
			],
		};

		await build({
			...structuredClone(defaultBuidConfig),
			plugins: [entriesPlugin(buildConfig)],
		});

		expect(readFile("fixtures/dist/index-06b5571c.es.js")).toMatchSnapshot();
		expect(readFile("fixtures/dist/other-ca5d566e.es.js")).toMatchSnapshot();
		expect(readFile("fixtures/dist/cjs/index-6fade66e.cjs.js")).toMatchSnapshot();
		expect(readFile("fixtures/dist/cjs/other-8548cf46.cjs.js")).toMatchSnapshot();

		expect(JSON.parse(readFile("fixtures/dist/cjs/package.json"))).toEqual(
			expect.objectContaining({
				type: "commonjs",
			}),
		);

		expect(JSON.parse(readFile("fixtures/package.json"))).toEqual(
			expect.objectContaining({
				exports: {
					"./index": {
						import: "./dist/index-06b5571c.es.js",
						require: "./dist/cjs/index-6fade66e.cjs.js",
						default: "./dist/index-06b5571c.es.js",
					},
					"./other": {
						import: "./dist/other-ca5d566e.es.js",
						require: "./dist/cjs/other-8548cf46.cjs.js",
						default: "./dist/other-ca5d566e.es.js",
					},
				},
			}),
		);
	});

	test("creates main & module references for main exports", async () => {
		const buildConfig: PluginOptions = {
			formats: ["es", "cjs"],
			entries: [
				{ sourcePath: path.join(__dirname, "fixtures/index.ts"), exports: { isMain: true } },
				{ sourcePath: path.join(__dirname, "fixtures/other.ts") },
			],
		};

		await build({
			...structuredClone(defaultBuidConfig),
			plugins: [entriesPlugin(buildConfig)],
		});

		expect(JSON.parse(readFile("fixtures/package.json"))).toEqual(
			expect.objectContaining({
				module: "./dist/index-06b5571c.es.js",
				main: "./dist/cjs/index-6fade66e.cjs.js",
			}),
		);
	});

	test("supports modifying output and exports paths", async () => {
		const buildConfig: PluginOptions = {
			formats: ["es", "cjs"],
			entries: [
				{
					sourcePath: path.join(__dirname, "fixtures/index.ts"),
					outputPath: "nestedFolder/helloWorld",
				},
				{
					sourcePath: path.join(__dirname, "fixtures/other.ts"),
					exports: { exportPath: "virtualFolder/bestGreeting" },
				},
			],
		};

		await build({
			...structuredClone(defaultBuidConfig),
			plugins: [entriesPlugin(buildConfig)],
		});

		expect(readFile("fixtures/dist/nestedFolder/helloWorld-06b5571c.es.js")).toMatchSnapshot();
		expect(readFile("fixtures/dist/cjs/nestedFolder/helloWorld-6fade66e.cjs.js")).toMatchSnapshot();

		expect(JSON.parse(readFile("fixtures/package.json"))).toEqual(
			expect.objectContaining({
				exports: {
					"./nestedFolder/helloWorld": {
						import: "./dist/nestedFolder/helloWorld-06b5571c.es.js",
						require: "./dist/cjs/nestedFolder/helloWorld-6fade66e.cjs.js",
						default: "./dist/nestedFolder/helloWorld-06b5571c.es.js",
					},
					"./virtualFolder/bestGreeting": {
						import: "./dist/other-ca5d566e.es.js",
						require: "./dist/cjs/other-8548cf46.cjs.js",
						default: "./dist/other-ca5d566e.es.js",
					},
				},
			}),
		);
	});

	test("supports nesting folders outputting the same filename", async () => {
		const buildConfig: PluginOptions = {
			formats: ["es", "cjs"],
			entries: [
				{
					sourcePath: path.join(__dirname, "fixtures/index.ts"),
					exports: { exportPath: "main" },
				},
				{
					sourcePath: path.join(__dirname, "fixtures/nested/index.ts"),
					exports: { exportPath: "nested" },
				},
			],
		};

		await build({
			...structuredClone(defaultBuidConfig),
			plugins: [entriesPlugin(buildConfig)],
		});

		expect(readFile("fixtures/dist/index-06b5571c.es.js")).toMatchSnapshot();
		expect(readFile("fixtures/dist/nested/index-c03f850b.es.js")).toMatchSnapshot();

		expect(readFile("fixtures/dist/cjs/index-6fade66e.cjs.js")).toMatchSnapshot();
		expect(readFile("fixtures/dist/cjs/nested/index-8dadecae.cjs.js")).toMatchSnapshot();
	});

	test.each(["main", "module", "exports"])(
		'fails if the existing package.json already contains the key %s that would be overwritten but #exports: "Generated automatically" is not found',
		async (key) => {
			const buildConfig: PluginOptions = {
				formats: ["es"],
				entries: [
					{
						sourcePath: path.join(__dirname, "fixtures/index.ts"),
					},
				],
			};

			fs.writeFileSync(
				path.join(__dirname, "fixtures/package.json"),
				JSON.stringify({ name: "fixtures", [key]: true }, undefined, 2),
			);

			await expect(async () => {
				await build({
					...structuredClone(defaultBuidConfig),
					plugins: [entriesPlugin(buildConfig)],
				});
			}).rejects.toThrowError(
				`Detected that the keys: "${key}" already exist in the package.json but the auto-generated marker "#exports" does not. Aborting in order not to overwrite these keys - add an "#exports": "Generated automatically" anywhere in your package.json to confirm these keys can be overwritten`,
			);
		},
	);
});
