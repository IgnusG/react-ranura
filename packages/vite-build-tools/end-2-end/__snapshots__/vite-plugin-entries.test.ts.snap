// Vitest Snapshot v1

exports[`vite-plugin-entries > correctly creates the output build files for a es/cjs mixed configuration 1`] = `
"const greeting = \\"Hello World!\\";
export {
  greeting
};
"
`;

exports[`vite-plugin-entries > correctly creates the output build files for a es/cjs mixed configuration 2`] = `
"\\"use strict\\";
Object.defineProperty(exports, Symbol.toStringTag, { value: \\"Module\\" });
const greeting = \\"Hello World!\\";
exports.greeting = greeting;
"
`;

exports[`vite-plugin-entries > correctly creates the output build files for a es/cjs mixed configuration for multiple entries 1`] = `
"const greeting = \\"Hello World!\\";
export {
  greeting
};
"
`;

exports[`vite-plugin-entries > correctly creates the output build files for a es/cjs mixed configuration for multiple entries 2`] = `
"const otherGreeting = \\"Hello everyone!\\";
export {
  otherGreeting
};
"
`;

exports[`vite-plugin-entries > correctly creates the output build files for a es/cjs mixed configuration for multiple entries 3`] = `
"\\"use strict\\";
Object.defineProperty(exports, Symbol.toStringTag, { value: \\"Module\\" });
const greeting = \\"Hello World!\\";
exports.greeting = greeting;
"
`;

exports[`vite-plugin-entries > correctly creates the output build files for a es/cjs mixed configuration for multiple entries 4`] = `
"\\"use strict\\";
Object.defineProperty(exports, Symbol.toStringTag, { value: \\"Module\\" });
const otherGreeting = \\"Hello everyone!\\";
exports.otherGreeting = otherGreeting;
"
`;

exports[`vite-plugin-entries > correctly creates the output build files for a simple es configuration 1`] = `
"const greeting = \\"Hello World!\\";
export {
  greeting
};
"
`;

exports[`vite-plugin-entries > supports modifying output and exports paths 1`] = `
"const greeting = \\"Hello World!\\";
export {
  greeting
};
"
`;

exports[`vite-plugin-entries > supports modifying output and exports paths 2`] = `
"\\"use strict\\";
Object.defineProperty(exports, Symbol.toStringTag, { value: \\"Module\\" });
const greeting = \\"Hello World!\\";
exports.greeting = greeting;
"
`;

exports[`vite-plugin-entries > supports nesting folders outputting the same filename 1`] = `
"const greeting = \\"Hello World!\\";
export {
  greeting
};
"
`;

exports[`vite-plugin-entries > supports nesting folders outputting the same filename 2`] = `
"const greeting = \\"Nested Hello World!\\";
export {
  greeting
};
"
`;

exports[`vite-plugin-entries > supports nesting folders outputting the same filename 3`] = `
"\\"use strict\\";
Object.defineProperty(exports, Symbol.toStringTag, { value: \\"Module\\" });
const greeting = \\"Hello World!\\";
exports.greeting = greeting;
"
`;

exports[`vite-plugin-entries > supports nesting folders outputting the same filename 4`] = `
"\\"use strict\\";
Object.defineProperty(exports, Symbol.toStringTag, { value: \\"Module\\" });
const greeting = \\"Nested Hello World!\\";
exports.greeting = greeting;
"
`;
